ChainReaders now apply their own transformations

mnmelo opened the pull request:
Closes #3343

Takes into account over-transformation of `SingleFrameReaders`, which
never re-create their `ts` when re-accessed.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

mnmelo committed:
ChainReaders now apply their own transformations

Closes #3343

Takes into account over-transformation of SingleFrameReaders, which
never re-create their ts when re-accessed.

pep8speaks commented:
Hello @mnmelo! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`package/MDAnalysis/coordinates/chain.py`](https://github.com/MDAnalysis/mdanalysis/blob/02a8ab8ca95c48a68e8db93ea2edab378888b847/package/MDAnalysis/coordinates/chain.py):

> [Line 580:71](https://github.com/MDAnalysis/mdanalysis/blob/02a8ab8ca95c48a68e8db93ea2edab378888b847/package/MDAnalysis/coordinates/chain.py#L580): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

* In the file [`testsuite/MDAnalysisTests/coordinates/test_chainreader.py`](https://github.com/MDAnalysis/mdanalysis/blob/02a8ab8ca95c48a68e8db93ea2edab378888b847/testsuite/MDAnalysisTests/coordinates/test_chainreader.py):

> [Line 182:13](https://github.com/MDAnalysis/mdanalysis/blob/02a8ab8ca95c48a68e8db93ea2edab378888b847/testsuite/MDAnalysisTests/coordinates/test_chainreader.py#L182): [E306](https://duckduckgo.com/?q=pep8%20E306) expected 1 blank line before a nested definition, found 0

mnmelo commented:
To prevent over-transformation of `SingleFrameReaders` this now executes an `active_reader.rewind()` if the `active_reader` is a `SingleFrameReader`. It works, but might penalize performance because the reader will read the Timestep at least twice.

I'm exploring the possibility that the ChainReader stores an unmodified Timestep along with the sub-reader.

mnmelo committed:
Also takes into account over-transformation of MemoryReaders

Overtransformation protection now implemented by book-keeping
already-transformed frames (for MemoryReaders and for SingleFrame
readers).

ChainReader now plays well with MemoryReader sub-readers. (Fixes #3349)

jbarnoud commented:
How does that work with the memory reader? This was the main reason the ChainReader is currently differing the transformations for the sub-readers.

mnmelo commented:
Good point. Will check. Tests passed, though this may just mean we need more tests for these conditions 😉

mnmelo commented:
I dug deeper and, from what I could understand, @jbarnoud, there should be no conflict with the MemoryReader. Here's the several cases that I see may happen when loading a Universe with a ChainReader:
1. `in_memory` requested at Universe creation time and `transformations` passed: in this case coordinates are first read into memory, and only then transformations added and applied. It will always be the MemoryReader (which has, by then, replaced the ChainReader) that will apply them, and this shouldn't differ from current behavior;
2. `transfer_to_memory` issued after the Universe is loaded with transformations: in this case timesteps will be transformed by the ChainReader, then read into memory. This seems to also be the current behavior.
3. `add_transformations` issued after the trajectory is in memory: this is essentially the same order of events as during Universe initialization.

Did you have any specific conflicting use-case in mind that I can double check? Perhaps there may be contrived corner cases where one of a ChainReader's sub-readers is a MemoryReader (is this even possible?). Was this what you were thinking about?

jbarnoud commented:
The case that I expect to break is if you have a memory reader within the chain reader. That could happen if one of the trajectories in the chain is an array of coordinates. Then you have a special case with the chain reader and a special case for the memory reader, question is : how do these special case work with each other? On 7 Jun 2021 02:51, Manuel Nuno Melo ***@***.***> wrote:
I dug deeper and, from what I could understand, @jbarnoud, there should be no conflict with the MemoryReader. Here's the several cases that I see may happen when loading a Universe with a ChainReader:
in_memory requested at Universe creation time and transformations passed: in this case coordinates are first read into memory, and only then transformations added and applied. It will always be the MemoryReader (which has, by then, replaced the ChainReader) that will apply them, and this shouldn't differ from current behavior;transfer_to_memory issued after the Universe is loaded with transformations: in this case timesteps will be transformed by the ChainReader, then read into memory. This seems to also be the current behavior.add_transformations issued after the trajectory is in memory: this is essentially the same order of events as during Universe initialization.
Did you have any specific conflicting use-case in mind that I can double check? Perhaps there may be contrived corner cases where one of a ChainReader's sub-readers is a MemoryReader (is this even possible?). Was this what you were thinking about?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

mnmelo commented:
Ok, so maybe not so much a corner case.

In that situation we have indeed a problem, in that the ChainReader risks applying the same transform multiple times, cumulatively, on the MemoryReader. I'm trying to make the ChainReader aware of this and somehow keep track of the transformed sub-frames.

mnmelo committed:
Extend docs to _apply_transformations

mnmelo commented:
The tests are passing, but there are some problems connecting to Codecov from GH Actions CI (a consequence of the fastly crash?). I keep re-running the GH Actions CI tests but there's always one or two that can't complete that post-test step. Would a rerun of all CI help? (how can I even trigger it?)

IAlibay commented:
> The tests are passing, but there are some problems connecting to Codecov from GH Actions CI (a consequence of the fastly crash?). I keep re-running the GH Actions CI tests but there's always one or two that can't complete that post-test step. Would a rerun of all CI help? (how can I even trigger it?)

@mnmelo I didn't get to see what happened. Maybe let's see what happens when this current round finishes running? There's a few options to re-run all the CI (the easiest is just to close and re-open the PR), but it's not necessary at the moment.

IAlibay commented:
Seemed to have worked this time. We can ignore the codecov project report (since it's just affected by the reduced number of lines).

orbeckst committed:
Merge branch 'develop' into issue-3343-chaintransformations

orbeckst commented:
Given the slew of ChainReader-related issues, I'd like to breath some new life into this PR...
